content/guides/additional/client.md

```diff
@@ -0,0 +1,2154 @@
+    SWOOSHING.src = "flap/audio/sfx-swooshing/wav";
+
+    const DIE = new Audio();
+    DIE.src = "flap/audio/sfx-die/wav";
+    ```
+
+Flappy Bird in HTML+JS needs Urbit affordances so it knows how to talk to the app backend.  Changes like the above will be common in adapting non-Hoon-compliant code due to assumptions about file location and naming conventions, but most of the time your process of adaptation will look much like the above.
+
+You can check out the API endpoints listed in the code above:  `/apps/flap/whoami`, `/apps/flap/img/sprite/png` and so forth, to see what they serve to the browser as.
+
+This version of the app should run in your browser correctly once you `|commit %flap`.  Urbit knows where all resources are and how to serve them.  However, while the vertical communications work (between client and server), the horizontal communications are still missing (between Urbit peers).
+
+
+##  Communications Protocol
+
+### Changes to Back End
+
+#### Urbit as Database
+
+Our app's state should retain a list of _all_ high scores it receives.  To that end, we will transition from a single high-score state to using a `(map fren score)`.
+
+```hoon
+/-  *flap, pals
+::  * * *
+::
+|%
++$  versioned-state
+  $%  state-zero
+      state-one
+  ==
++$  state-one  $:
+      %one
+      =score
+      =scores
+    ==
++$  state-zero  $:
+      %zero
+      =score
+      hiscore=score
+    ==
++$  card  card:agent:gall
+--
+```
+
+We will include an upgrade path for the old agent:
+
+```hoon
+++  on-load
+  |=  old-state=vase
+  ^-  (quip card _this)
+  =/  old  !<(versioned-state old-state)
+  ?-  -.old
+    %one   `this(state old)
+    %zero  `this(scores (~(put by scores) our.bol hiscore.old))
+  ==
+```
+
+These values will be changed when peer state changes are received using `%pals`.
+
+#### Adding Friends
+
+`%pals` is a very simple contact manager which recognizes outgoing requests, incoming requests, and mutually recognized peers.  Let's integrate knowledge of other friends (and thus the ability to maintain a leaderboard of our friends).  We will base our leaderboard on outgoing requests for simplicity.
+
+You should copy `%pals` support files over to our working desk.
+
+```sh
+# Include the %pals tooling by ~paldev.
+git clone https://github.com/Fang-/suite.git
+cp suite/lib/pals.hoon comet/flap/lib
+cp suite/sur/pals.hoon comet/flap/sur
+cp -r suite/mar/pals comet/flap/mar
+```
+
+When dealing with `%pals`, we need to maintain some list of our friends and their current known high scores (which means modifying the app's state) and dealing with sending and receiving notifications.  That is reflected in adding `scores` to the new state in the state definition in `/app/flap.hoon`.
+
+Subscriptions will all take place over the `/flap` path.  In the `++on-poke` arm, we need to issue notices along that arm with the `%flap-update` mark.
+
+```hoon
+  %flap-action
+=/  axn  !<(action vase)
+?>  =(-.axn %gain)
```

Comment:
this and similar isntances should really be a `?=` not a `.=`. If it's working currently it's cos the tagged union only contains a single entry and the compiler is maybe smart about it or reducing it. If there were others in the tagged union it would not be able to find the face

Reply:
yep, good catch, fixed